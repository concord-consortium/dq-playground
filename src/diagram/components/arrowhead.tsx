import React from "react";
import { Position } from "reactflow";

export const kArrowheadSize = 10;

interface IArrowheadProps {
  color?: string;
  targetPosition: Position;
  targetX: number;
  targetY: number;
}
export const Arrowhead = ({ color, targetPosition, targetX, targetY }: IArrowheadProps) => {
  const _color = color || "#ff0000";
  const width = 10;
  const height = 10;

  const xOffset1 = targetPosition === Position.Left ? -width
    : targetPosition === Position.Right ? width
    : -width / 2;
  const yOffset1 = targetPosition === Position.Top ? -height
    : targetPosition === Position.Bottom ? height
    : -height / 2;
  const xOffset3 = [Position.Left, Position.Right].includes(targetPosition) ? xOffset1 : -xOffset1;
  const yOffset3 = [Position.Top, Position.Bottom].includes(targetPosition) ? yOffset1 : -yOffset1;
  const [ahx1, ahy1] = [targetX + xOffset1, targetY + yOffset1];
  const [ahx2, ahy2] = [targetX, targetY];
  const [ahx3, ahy3] = [targetX + xOffset3, targetY + yOffset3];
  const arrowheadPoints = `${ahx1},${ahy1} ${ahx2},${ahy2} ${ahx3},${ahy3} ${ahx1},${ahy1}`;

  return (
    <polygon
      // data-testid="marker-end-polygon"
      fill={_color}
      points={arrowheadPoints}
      stroke={_color}
      strokeLinecap="square"
      strokeLinejoin="miter"
      strokeWidth="2.5"
      x={targetX}
      y={targetY}
    />
  );
};
