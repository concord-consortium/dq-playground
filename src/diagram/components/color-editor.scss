@import "../../scss/vars.scss";
@import "../../scss/mixins.scss";

.color-editor-dialog {

  &.blue { @include spread-map($theme-map-blue); }
  &.gray { @include spread-map($theme-map-gray); }
  &.green { @include spread-map($theme-map-green); }
  &.light-gray { @include spread-map($theme-map-light-gray); }
  &.red { @include spread-map($theme-map-red); }
  &.yellow { @include spread-map($theme-map-yellow); }

  align-items: center;
  background-color: theme-var($--theme-color-5);
  border: solid 2px theme-var($--theme-color-2);
  border-radius: 5px;
  border-top-left-radius: 0;
  box-sizing: border-box;
  display: flex;
  height: 30px;
  left: -119px;
  padding: 5px 0 6px 8px;
  position: absolute;
  top: 22px;
  width: 143px;
  z-index: 1;

  div[title="#{$color-blue-5}"] {
    border: solid 1px $color-blue-2;
  }
  div[title="#{$color-gray-5}"] {
    border: solid 1px $color-gray-2;
  }
  div[title="#{$color-light-gray-5}"] {
    border: solid 1px $color-light-gray-2;
  }
  div[title="#{$color-green-5}"] {
    border: solid 1px $color-green-2;
  }
  div[title="#{$color-red-5}"] {
    border: solid 1px $color-red-2;
  }
  div[title="#{$color-yellow-5}"] {
    border: solid 1px $color-yellow-2;
  }
}
