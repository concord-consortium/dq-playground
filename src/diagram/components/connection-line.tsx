import React from "react";
import { ConnectionLineComponent, ConnectionLineComponentProps } from "reactflow";

export const ConnectionLine: ConnectionLineComponent = ({ fromX, fromY, toX, toY, ...rest }: ConnectionLineComponentProps) =>  {
  console.log(`connecstionLine`, {fromX, fromY, toX, toY, rest});
  return (
    <g className="react-flow__connection">
      <path
        className="react-flow__connection-path"
        data-testid="connection-line"
        fill="none"
        stroke="#0081ff"
        strokeWidth={2}
        d={`M${fromX},${fromY} C ${fromX} ${toY} ${fromX} ${toY} ${toX},${toY}`}
      />
    </g>
  );
};
