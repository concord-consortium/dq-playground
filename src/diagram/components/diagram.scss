@import "../../scss/vars.scss";

.def-container {
  height: 0;
  width: 0;
}

// .react-flow__node.selected {
//   border: 1px solid black;
//   border-radius: 5px;
// }

.react-flow__connection:focus,
.react-flow__connection:focus * {
  outline: none;
}

// While some stroke settings affect the path markers (arrowheads), stroke color
// settings do not. To support different colored arrowheads, we define custom SVG
// markers in the Diagram component and set marker-end properties to the custom 
// markers' IDs in the style rules below. 
.react-flow__edge {
  &.used-in-expression {
    .react-flow__edge-path {
      marker-end: url("#custom-arrow__selected-or-used");
      stroke: $color-gray-1;
      stroke-dasharray: 0;
      stroke-width: 2;
    }
  }

  &.selected {
    .react-flow__edge-path {
      marker-end: url("#custom-arrow__selected-or-used");
    }
  }

  .react-flow__edge-path {
    marker-end: url("#custom-arrow");
    stroke: $color-light-gray-2;
    stroke-dasharray: 0, 10;
    stroke-linecap: round;
    stroke-width: 3;
  }
}

.react-flow__connection-path {
  marker-end: url("#custom-arrow__dragging");
  stroke: $select-blue;
  stroke-dasharray: 0, 10;
  stroke-linecap: round;
  stroke-width: 3;
}

.nested-set-node {
  display: inline-block;
  border: 1px solid black;
  padding: 5px;
}
