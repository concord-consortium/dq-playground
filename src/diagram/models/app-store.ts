import { types } from "mobx-state-tree";
import { DQRoot } from "..";
import { Variables } from "./variables";

export const AppStore = types.model("AppStore", {
  diagram: DQRoot,
  variables: Variables
});
