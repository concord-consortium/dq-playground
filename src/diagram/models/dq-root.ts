import { Instance, types } from "mobx-state-tree";
import { Elements, FlowTransform } from "react-flow-renderer/nocss";
import { DQNode, DQNodeType } from "./dq-node";
import { VariableType } from "./variable";

export interface VariablesAPI {
  createVariable: () => VariableType;
  removeVariable: (variable?: VariableType) => void;
}

export const DQRoot = types.model("DQRoot", {
  nodes: types.map(DQNode),
  flowTransform: types.maybe(types.frozen<FlowTransform>())
})
.volatile(self => ({
  variablesAPI: undefined as VariablesAPI | undefined,
  connectingVariable: undefined as VariableType | undefined
}))
.views(self => ({
  get reactFlowElements() {
    const elements: Elements = [];
    self.nodes.forEach((node) => {
      elements.push(...node.getReactFlowElements(self));
    });
    return elements;
  },
  get nodeFromVariableMap() {
    const map: Record<string, DQNodeType> = {};
    self.nodes.forEach((node) => {
      map[node.variableId] = node;
    });
    return map;
  }
}))
.views(self => ({
  getNodeFromVariableId(id: string) {
    return self.nodeFromVariableMap[id];
  }
}))
.actions(self => ({
  addNode(newNode: DQNodeType) {
    self.nodes.put(newNode);
  },
  removeNode(node: DQNodeType) {
    const variables = self.variablesAPI;
    if (!variables) {
      throw new Error("Need variables before deleting nodes");
    }

    variables.removeVariable(node.variable);
  },
  setTransform(transform: FlowTransform) {
    self.flowTransform = transform;
  }
}))
.actions(self => ({
  createNode({x,y}: {x:number, y:number}) {
    const variables = self.variablesAPI;
    if (!variables) {
      throw new Error("Need variables before creating nodes");
    }
    const variable = variables.createVariable();
    const node = DQNode.create({variable: variable.id, x, y});
    self.addNode(node);
  },

  setVariablesAPI(variablesAPI: VariablesAPI) {
    self.variablesAPI = variablesAPI;
  },
  setConnectingVariable(variable?: VariableType) {
    self.connectingVariable = variable;
  }
}));
export interface DQRootType extends Instance<typeof DQRoot> {}
