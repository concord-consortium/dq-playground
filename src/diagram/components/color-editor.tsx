import React, { /*useEffect, useRef,*/ useState } from "react";
//import { useClickAway } from "../../hooks/use-click-away";
import { VariableType } from "../models/variable";

import "./color-editor.scss";
import {CirclePicker} from "react-color";

interface IProps {
  variable: VariableType
  onShowColorEditor: (show: boolean) => void;
}

export const ColorEditor: React.FC<IProps> = ({variable, onShowColorEditor}) => {

  const [color, setColor] = useState(variable.color);

  const handleColorChange = (evt:any) => {
    variable.setColor(evt.hex);
    setColor(evt.hex);
    onShowColorEditor(false);
  };

  return (
    <div className="color-editor-dialog" title={"color picker"}>
      <div className="variable-info-row ">
        <CirclePicker color={color} onChange={handleColorChange} width={"110px"} colors={
          [
            "#FF6900",
            "#e98b42",
            "#7BDCB5",
            "#00D084",
            "#8ED1FC",
            "#0693E3",
            "#ABB8C3",
            "#EB144C",
            "#F78DA7",
            "#9900EF"
          ]} circleSize={14} circleSpacing={7}/>
      </div>
    </div>
  );
};
