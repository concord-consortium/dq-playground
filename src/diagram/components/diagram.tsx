import { observer } from "mobx-react-lite";
import React, { useEffect, useRef, useState } from "react";
import ReactFlow, { Edge, Elements, OnConnectFunc, isEdge,
  OnEdgeUpdateFunc, MiniMap, Controls, ReactFlowProvider, FlowTransform, OnConnectStartFunc, OnConnectEndFunc } from "react-flow-renderer/nocss";
import { DQRootType } from "../models/dq-root";
import { DQNodeType } from "../models/dq-node";
import { QuantityNode } from "./quantity-node";
import { FloatingEdge } from "./floating-edge";
import { ToolBar } from "./toolbar";

// We use the nocss version of RF so we can manually load
// the CSS. This way we can override it.
// Otherwise RF injects its CSS after our CSS, so we can't
// override it.
import "react-flow-renderer/dist/style.css";
import "react-flow-renderer/dist/theme-default.css";

// The order matters the diagram css overrides some styles
// from the react-flow css.
import "./diagram.scss";

const nodeTypes = {
  quantityNode: QuantityNode,
};
const edgeTypes = {
  floatingEdge: FloatingEdge,
};

interface IProps {
  dqRoot: DQRootType;
  showEditVariableDialog?: () => void;
  showNestedSet?: boolean;
  getDiagramExport?: () => unknown;
}
export const _Diagram = ({ dqRoot, getDiagramExport, showEditVariableDialog }: IProps) => {
  const reactFlowWrapper = useRef<any>(null);
  const [selectedNode, setSelectedNode] = useState<DQNodeType | undefined>();
  const [rfInstance, setRfInstance] = useState<any>();

  // Keep the model's selected node up to date with ReactFlow's
  useEffect(() => {
    dqRoot.setSelectedNode(selectedNode);
  }, [dqRoot, selectedNode]);

  const handleChangeFlowTransform = (transform?: FlowTransform) => {
    transform && dqRoot.setTransform(transform);
  };

  const onConnectStart: OnConnectStartFunc = (event, { nodeId, handleType }) => {
    if (!nodeId) {
      return;
    }
    const node = dqRoot.getNodeFromVariableId(nodeId);
    dqRoot.setConnectingVariable(node.variable);
  };
  const onConnectEnd: OnConnectEndFunc = () => {
    dqRoot.setConnectingVariable(undefined);
  };

  // gets called after end of edge gets dragged to another source or target
  const onEdgeUpdate: OnEdgeUpdateFunc = (oldEdge, newConnection) => {
    // We could try to be smart about this, and only update things that
    // changed but it is easier to just break the old connection
    // and make a new one
    const oldTargetNode = dqRoot.getNodeFromVariableId(oldEdge.target);
    const oldSourceNode = dqRoot.getNodeFromVariableId(oldEdge.source);
    oldTargetNode?.removeInput(oldSourceNode.variable);
    const { source, target } = newConnection;
    const newSourceNode = source ? dqRoot.getNodeFromVariableId(source) : undefined;
    const newTargetNode = target ? dqRoot.getNodeFromVariableId(target) : undefined;
    newTargetNode?.addInput(newSourceNode);
  };

  const onConnect: OnConnectFunc = (params) => {
    const { source, target, targetHandle } = params;
    console.log(`connection source: ${source} to target: ${target} on handle: ${targetHandle}`);
    if ( source && target ) {
      const targetModel = dqRoot.getNodeFromVariableId(target);
      const sourceModel = dqRoot.getNodeFromVariableId(source);
      targetModel?.addInput(sourceModel);
    }
  };

  const onElementsRemove = (elementsToRemove: Elements) => {
    for(const element of elementsToRemove) {
      if (isEdge((element as any))) {
        const edge = element as Edge;
        const { source, target } = edge;
        const targetModel = dqRoot.getNodeFromVariableId(target);
        const sourceModel = dqRoot.getNodeFromVariableId(source);
        // sourceModel gets removed first when a node is selected to be deleted, otherwise, just remove the connection
        sourceModel.tryVariable && targetModel.removeInput(sourceModel.variable);
      } else {
        // If this is the selected node we need to remove it from the state too
        const nodeToRemove = dqRoot.getNodeFromVariableId(element.id);
        setSelectedNode((currentNode) => nodeToRemove === currentNode ? undefined : currentNode);
        dqRoot.removeNode(nodeToRemove);
      }
    }
  };

  const onSelectionChange = (selectedElements: Elements | null) => {
    if (selectedElements?.[0]?.type === "quantityNode" ) {
      setSelectedNode(dqRoot.getNodeFromVariableId(selectedElements[0].id));
    } else {
      setSelectedNode(undefined);
    }
  };

  const onDragOver = (event: any) => {
    event.preventDefault();
    event.dataTransfer.dropEffect = "move";
  };

  const onDrop = (event: any) => {
    event.preventDefault();

    if (!reactFlowWrapper.current || !rfInstance) {
      return;
    }

    const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
    const position = rfInstance.project({
      x: event.clientX - reactFlowBounds.left,
      y: event.clientY - reactFlowBounds.top,
    });

    dqRoot.createNode({
      x: position.x,
      y: position.y
    });
  };

  const onNodeDrag = (event: React.MouseEvent<Element, MouseEvent>, node: any) => {
    event.stopPropagation();
  };

  // Keep the MST node model in sync with the diagram
  const onNodeDragStop = (event: any, node: any) => {
    const mstNode = dqRoot.getNodeFromVariableId(node.id);
    mstNode?.updatePosition(node.position.x, node.position.y);
    event.stopPropagation();
  };

  const { zoom: defaultZoom, x, y } = dqRoot.flowTransform || {};
  const defaultPosition: [number, number] | undefined = x != null && y != null ? [x, y] : undefined;

  return (
    <div className="diagram" ref={reactFlowWrapper} data-testid="diagram">
      <ReactFlowProvider>
        <ReactFlow
          elements={dqRoot.reactFlowElements}
          defaultPosition={defaultPosition}
          defaultZoom={defaultZoom}
          nodeTypes={nodeTypes}
          edgeTypes={edgeTypes}
          onEdgeUpdate={onEdgeUpdate}
          onConnect={onConnect as any}  // TODO: fix types
          onConnectStart={onConnectStart}
          onConnectEnd={onConnectEnd}
          onElementsRemove={onElementsRemove}
          onSelectionChange={onSelectionChange}
          onLoad={(_rfInstance) => setRfInstance(_rfInstance)}
          onDrop={onDrop}
          onDragOver={onDragOver}
          onNodeDrag={onNodeDrag}
          onNodeDragStop={onNodeDragStop}
          onMoveEnd={handleChangeFlowTransform}>
          <MiniMap/>
          <Controls />
          <ToolBar {...{dqRoot, getDiagramExport, showEditVariableDialog}}/>
        </ReactFlow>
      </ReactFlowProvider>
    </div>
  );
};

export const Diagram = observer(_Diagram);
Diagram.displayName = "Diagram";
