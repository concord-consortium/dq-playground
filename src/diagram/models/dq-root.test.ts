import { AppStore } from "../../components/app-store";
import defaultDiagram from "../../components/default-diagram";
import { DQRoot } from "./dq-root";

describe("DQRoot", () => {
  it("can be created", () => {
    const root = DQRoot.create();
    expect(root.nodes).toBeDefined();
  });

  it("can generate react flow nodes", () => {
    const appStore = AppStore.create(defaultDiagram);
    const root = appStore.diagram;
    const nodes = root.reactFlowNodes;
    const readOnlyNodes = root.reactFlowNodesReadOnly;

    expect({ nodes, readOnlyNodes }).toMatchSnapshot();
  });

  it("can generate react flow edges", () => {
    const appStore = AppStore.create(defaultDiagram);
    const root = appStore.diagram;
    const edges = root.reactFlowEdges;

    expect(edges).toMatchSnapshot();
  });
});
