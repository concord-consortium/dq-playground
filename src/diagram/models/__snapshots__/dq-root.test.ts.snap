// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DQRoot can generate react flow elements 1`] = `
Array [
  Object {
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
      "node": Object {
        "id": "1",
        "variable": "a",
        "x": 100,
        "y": 100,
      },
    },
    "id": "a",
    "position": Object {
      "x": 100,
      "y": 100,
    },
    "type": "quantityNode",
  },
  Object {
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
      "node": Object {
        "id": "2",
        "variable": "b",
        "x": 100,
        "y": 300,
      },
    },
    "id": "b",
    "position": Object {
      "x": 100,
      "y": 300,
    },
    "type": "quantityNode",
  },
  Object {
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
      "node": Object {
        "id": "3",
        "variable": "c",
        "x": 550,
        "y": 150,
      },
    },
    "id": "c",
    "position": Object {
      "x": 550,
      "y": 150,
    },
    "type": "quantityNode",
  },
  Object {
    "arrowHeadType": "arrowclosed",
    "className": "",
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
    },
    "id": "ea-targetc-a",
    "source": "a",
    "target": "c",
    "type": "floatingEdge",
  },
  Object {
    "arrowHeadType": "arrowclosed",
    "className": "",
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
    },
    "id": "ed-targetc-a",
    "source": "d",
    "target": "c",
    "type": "floatingEdge",
  },
  Object {
    "arrowHeadType": "arrowclosed",
    "className": "",
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
    },
    "id": "eb-targetc-a",
    "source": "b",
    "target": "c",
    "type": "floatingEdge",
  },
  Object {
    "data": Object {
      "dqRoot": Object {
        "flowTransform": undefined,
        "nodes": Object {
          "1": Object {
            "id": "1",
            "variable": "a",
            "x": 100,
            "y": 100,
          },
          "2": Object {
            "id": "2",
            "variable": "b",
            "x": 100,
            "y": 300,
          },
          "3": Object {
            "id": "3",
            "variable": "c",
            "x": 550,
            "y": 150,
          },
          "4": Object {
            "id": "4",
            "variable": "d",
            "x": 150,
            "y": 500,
          },
        },
      },
      "node": Object {
        "id": "4",
        "variable": "d",
        "x": 150,
        "y": 500,
      },
    },
    "id": "d",
    "position": Object {
      "x": 150,
      "y": 500,
    },
    "type": "quantityNode",
  },
]
`;
